### PR TITLE
refactor: re-organize top level modules

### PR DESCRIPTION
**Pull request by roeap**
# Description

~This contains changes from #1432, will rebase once that's merged.~

This PR constitutes the bulk of re-organising our top level modules.
- move `DeltaTable*` structs into new `table` module
- move table configuration into `table` module
- move schema related modules into `schema` module
- rename `action` module to `protocol` - hoping to isolate everything that can one day be the log kernel.

~It also removes the deprecated commit logic from `DeltaTable`  and updates call sites and tests accordingly.~

I am planning one more follow up, where I hope to make `transactions` currently within `operations` a top level module. While the number of touched files here is already massive, I want to do this in a follow up, as it will also include some updates to the transactions itself, that should be more carefully reviewed.

# Related Issue(s)

closes: #1136

# Documentation

### CODE REVIEW

File: python/src/error.rs [hunk 1]
```diff
@@ -63,11 +63,14 @@ fn checkpoint_to_py(err: ProtocolError) -> PyErr {
     match err {
         ProtocolError::Arrow { source } => arrow_to_py(source),
         ProtocolError::ObjectStore { source } => object_store_to_py(source),
+        ProtocolError::EndOfLog => DeltaProtocolError::new_err("Enf of log"),
```
Comment (cmackenzie1) [inline python/src/error.rs:66]:
```suggestion
        ProtocolError::EndOfLog => DeltaProtocolError::new_err("End of log"),
```

File: rust/src/operations/transaction/mod.rs [hunk 2]
```diff
@@ -164,7 +164,8 @@ pub async fn commit(
 ) -> DeltaResult<i64> {
     let tmp_commit = prepare_commit(storage, &operation, actions, app_metadata).await?;
 
-    let max_attempts = 5;
+    // TODO make max attempts configurable
+    let max_attempts = 25;
```
Comment (roeap) [inline rust/src/operations/transaction/mod.rs:168]: Increasing this was needed to make our concurrent writer tests pass.

File: rust/tests/integration_commit.rs
Comment (roeap) [inline rust/tests/integration_commit.rs:1]: these tests we just covering the old commit logic.

File: rust/src/operations/transaction/mod.rs (hunk 2 above)
Comment (wjones127) [reply rust/src/operations/transaction/mod.rs:168]: Huh. I guess 5 workers trying to make 3 commits. That mean one potentially has to fail 14 times?

**Review by wjones127 — APPROVED** [review]
This looks good. I like that there aren't as many top level modules in the end. We should do a round of Rust documentation improvements after this.

I'd like @rtyler's feedback before we merge though.

File: rust/src/operations/transaction/mod.rs (hunk 2 above)
Comment (roeap) [reply rust/src/operations/transaction/mod.rs:168]: 10 was not cutting it for s3, so at least that...

**Comment by roeap** [comment]
> We should do a round of Rust documentation improvements after this

Was thinking the same. At the very lest we should be able to say what each module does ...

**Comment by roeap** [comment]
@rtyler - I'd love to get this merged, as I have some work lined up dependent on this. Before we'd love to hear your opinion.

The number of changed files may look daunting, but its mostly renames 😉.

**Comment by roeap** [comment]
@rtyler @wjones127 - any chance we can get this merged?

IF you feel this should be abandoned, let me know as well, then I'll stop using this as a baseline for follow on work 😆.

**Comment by ion-elgreco** [comment]
> I I would really like you to do some refactoring. I'm currently working on implementing CDF, but I feel that if the source code is more structured with clearer guidelines, it will be easier to determine where to implement it.

How is adding CDF going? :D Would be awesome to have

**Review by wjones127 — APPROVED** [review]

**Pull request merged**